File: engine/julia/strategic_gamma_scalping_engine.jl
# Module AF5 (Julia): Strategic Gamma Scalping & Position Adjustment Engine
# Precision calculation of Gamma Decay Breakeven ("Paying the Rent") and daily sigma moves

module StrategicGammaScalpingEngine

export calculate_gamma_decay_breakeven, calculate_daily_sigma_move

function calculate_gamma_decay_breakeven(daily_theta::Float64, position_gamma::Float64)
    g = max(1e-6, position_gamma)
    th = abs(daily_theta)
    return round(sqrt((2.0 * th) / g), digits=4)
end

function calculate_daily_sigma_move(spot::Float64, annual_vol::Float64)
    daily_vol = annual_vol / sqrt(252.0)
    sigma1 = spot * daily_vol
    return (
        daily_vol_pct = round(daily_vol * 100.0, digits=4),
        one_sigma_move = round(sigma1, digits=2),
        upper_1sigma = round(spot + sigma1, digits=2),
        lower_1sigma = round(spot - sigma1, digits=2)
    )
end

end # module

File: engine/rust/src/strategic_gamma_scalping_engine.rs
//! Module AF4 (Rust): Strategic Gamma Scalping & Position Adjustment Engine.
//! High-speed gamma scalping rebalancer, gamma decay breakeven calculator, and spread roller.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct StrategicGammaScalpingState {
    pub spot_price: f32,
    pub last_hedge_spot: f32,
    pub position_gamma: f32,
    pub daily_theta_rent: f32,
    pub gamma_decay_breakeven: f32,
    pub daily_one_sigma_move: f32,
    pub net_delta: f32,
    pub rebalance_shares_needed: i32,
    pub roll_operation_id: u32,
    pub staying_alive_flag: u32,
    pub _padding: [u8; 24],
}

pub struct StrategicGammaScalpingEngine;

impl StrategicGammaScalpingEngine {
    pub fn new_state() -> StrategicGammaScalpingState {
        StrategicGammaScalpingState {
            spot_price: 100.0,
            last_hedge_spot: 100.0,
            position_gamma: 0.15,
            daily_theta_rent: 0.03,
            gamma_decay_breakeven: 0.632,
            daily_one_sigma_move: 1.57,
            net_delta: 0.0,
            rebalance_shares_needed: 0,
            roll_operation_id: 1,
            staying_alive_flag: 1,
            _padding: [0; 24],
        }
    }

    pub fn execute_scalp_evaluation(
        state: &mut StrategicGammaScalpingState,
        spot: f32, last_hedge: f32,
        gamma: f32, theta: f32, net_delta: f32,
        annual_vol: f32,
    ) {
        state.spot_price = spot;
        state.last_hedge_spot = last_hedge;
        state.position_gamma = gamma.max(1e-6);
        state.daily_theta_rent = theta.abs();
        state.net_delta = net_delta;

        state.gamma_decay_breakeven = ((2.0 * state.daily_theta_rent) / state.position_gamma).sqrt();
        let daily_vol = annual_vol / (252.0_f32).sqrt();
        state.daily_one_sigma_move = spot * daily_vol;

        let move_dist = spot - last_hedge;
        state.rebalance_shares_needed = if move_dist.abs() >= 2.0 {
            -(net_delta * 100.0) as i32
        } else {
            0
        };

        state.staying_alive_flag = 1;
        state.roll_operation_id = 1;
    }
}
